File: Contests/AtCoder Beginner Contest 115/AtCoder Beginner Contest 115 - D.py
# AtCoder Beginner Contest 115 - D.py
anss = [0, 0, 0, 1, 2, 3, 3]
lenarr = [0,6]
phitab = [0,3]

def phi(i):
    return phitab[i]


def getlen(i):
    return lenarr[i]


def f(x, n):
    if n == 1:
        return anss[x]
    if x > getlen(n)/2:
        return phi(n) - f(getlen(n)-x+1, n)
    else:
        return f(x-1, n-1)

for i in range(2,60):
    lenarr.append(2*lenarr[i-1]+2)
for i in range(2,60):
    phitab.append(2*phitab[i-1]+1)
arg = input()
N = int(arg.split()[0])
X = int(arg.split()[1])
print(f(X+1, N))

File: Contests/AtCoder Beginner Contest 115/AtCoder Beginner Contest 115 - D.cpp
// AtCoder Beginner Contest 115 - D.cpp
#include <iostream>
#include <algorithm>
#include <cstdio>
#define ll long long
using namespace std;
int phi(int i)
{
    return (1 << (i + 1)) - 1;
}
int len(int i)
{
    return (1 << (i + 2)) - 2;
}
int anss[7] = {0, 0, 0, 1, 2, 3, 3};
int f(int x, int n)
{
    if (n == 1)
        return anss[x];
    if (x > len(n) / 2)
        return phi(n) - f(len(n) - x, n);
    else
        return f(x - 1, n - 1);
}

int main()
{
    int N, X;
    scanf("%d%d", &N, &X);
    cout << f(X + 1, N);
    return 0;
}